File: ch31/App_Code/VB/Listing31_2.vb
Imports Microsoft.VisualBasic
Imports System.Net

Public Class Listing31_2
    Public Sub RSSFeedLengthAsync(uris As IList(Of Uri))
        RSSFeedLengthAsyncHelper(uris.GetEnumerator(), 0)
    End Sub

    Private Sub RSSFeedLengthAsyncHelper(enumerator As IEnumerator(Of Uri), total As Integer)
        If enumerator.MoveNext() Then
            Dim client = New WebClient()
            AddHandler client.DownloadDataCompleted,
                Sub(sender, e)
                    RSSFeedLengthAsyncHelper(enumerator, total + e.Result.Length)
                End Sub
            client.DownloadDataAsync(enumerator.Current)
        Else
            enumerator.Dispose()
        End If
    End Sub
End Class

File: ch31/App_Code/CS/Listing31-2.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Web;

/// <summary>
/// Summary description for Listing02
/// </summary>
public class Listing31_2
{
	public Listing31_2()
	{
		
	}
    public void RSSFeedLengthAsync(IList<Uri> uris)
    {
        RSSFeedLengthAsyncHelper(uris.GetEnumerator(), 0);
    }

    private void RSSFeedLengthAsyncHelper(IEnumerator<Uri> enumerator, int total)
    {
        if (enumerator.MoveNext())
        {
            var client = new WebClient();
            client.DownloadDataCompleted += (sender, e) =>
            {
                RSSFeedLengthAsyncHelper(enumerator, total + e.Result.Length);
            };
            client.DownloadDataAsync(enumerator.Current);
        }
        else
        {
            enumerator.Dispose();
        }
    }
    
}